### PR TITLE
Implement `IntoResponse` for `Response<()>` and `response::Parts`

### PR DESCRIPTION
**Pull request by neoeinstein**
Add a simple implementation to use `Parts` and `Response<()>` as a source for `IntoResponse`.

This doesn't fit neatly into the `IntoResponseParts` paradigm because `Parts` already contains a status code. The implementation itself is rather trivial.

## Motivation

I have a library where I generate some generic template responses using `http::response::Builder` and wanted to be able to make it one step easier to use those templates (often built with a body of `()`, and then substituted with the intended body) as responses. This makes it so that the template can be built and specified next to the body to be placed into the response.

## Solution

Add impls for `Parts` and `Response<()>`. The new `Response<()>` impl discards the existing `()` body and uses the provided body to construct the response and then moves the headers and status into the new response.

### CODE REVIEW

**Comment by jplatte** [comment]
I think the `Parts` one might make sense, but

* it should be done in the macro below, in an equivalent manner to
  `impl<R, $($ty,)*> IntoResponse for (StatusCode, $($ty),*, R)`
* it should propagate response extensions as well

I think the `Response<X>` one is too much of a footgun though.

**Comment by neoeinstein** [comment]
Yep. The `Extensions` and `Version` were both things I looked at, but it's difficult to merge in the extensions as mentioned in that linked issue, as there's no way to iterate and extend at the moment.

I can see `Response<X>` being a potential footgun. It's a bit difficult because I can't provide an impl for `http::response::Response<R> where R: IntoResponse`, as that would conflict with the existing one that expects a fully formed `http_body::Body`. This was my workaround for that, but even just having `Parts` would make it one step easier.

File: axum-core/src/response/into_response.rs [hunk 1]
```diff
@@ -396,6 +396,27 @@ where
     }
 }
 
+impl<R> IntoResponse for (http::response::Parts, R)
+where
+    R: IntoResponse,
+{
+    fn into_response(self) -> Response {
+        let (parts, res) = self;
+        (parts.status, parts.headers, res).into_response()
```
Comment (davidpdrsn) [inline axum-core/src/response/into_response.rs:405]: Should we be replacing the version here? If not, why? The general pattern we try to follow is that things further to the left in the tuple takes precedence.

We did previously talk about `Version` and decided to just let hyper handle it or require the user to manually build a response. See https://github.com/tokio-rs/axum/issues/802.

I'm not sure what the right call is here.

Comment (jplatte) [reply axum-core/src/response/into_response.rs:405]: We could request that the http crate explicitly specifies that the default value is `HTTP_11` and add an assert for it 😄

Comment (neoeinstein) [reply axum-core/src/response/into_response.rs:405]: > I think we should probably also merge the headers with `HeaderMap::extend`.

`IntoResponseParts` does the right thing here. If there are any headers in the `HeaderMap`, then the headers from `R::into_response()` will be extended with those headers.

Comment (davidpdrsn) [reply axum-core/src/response/into_response.rs:405]: Ah yes I see thats right!

**Comment by davidpdrsn** [comment]
I think this looks good as is but I don't think we should merge it until https://github.com/hyperium/http/pull/546 has landed.

**Comment by davidpdrsn** [comment]
I think this looks good as is but I don't think we should merge it until https://github.com/hyperium/http/pull/546 has landed.

**Review by daheige — APPROVED** [review]

**Review by davidpdrsn — APPROVED** [review]
I think this is good to go now :)

**Review by jplatte — APPROVED** [review]
Changes look good either way.

**Comment by davidpdrsn** [comment]
Good idea! I filed https://github.com/tokio-rs/axum/issues/979. I'll look into it.

**Pull request merged**